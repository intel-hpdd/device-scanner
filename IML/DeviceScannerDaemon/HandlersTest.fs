module IML.DeviceScannerDaemon.HandlersTest

open IML.DeviceScannerDaemon.Handlers
open IML.DeviceScannerDaemon.TestFixtures
open Fable.Import.Node
open Fable.Core.JsInterop
open Fable.Core
open Fable.PowerPack
open Fable.Import
open Fable.Import.Jest
open Fable.Import.Jest.Matchers

let toJson =  Json.ofString >> Result.unwrapResult

testList "Read Handler" [
  let withSetup f () =
    let getUdevDB = Matcher<unit, JS.Promise<string>>(fun () ->
      Promise.lift ""
    )

    let ``end`` = Matcher<string option, unit>()

    promise {
      do! handleReadEvent' getUdevDB.Mock ``end``.Mock

      f (getUdevDB, ``end``)
    }

  yield! testFixtureAsync withSetup [
    "should call getUdevDB", fun (getUdevDB, ``end``) ->
      expect.Invoke(getUdevDB.Calls).toEqual([|[||]|])

    "should call end", fun (getUdevDB, ``end``) ->
      ``end`` <?> None
  ]
]

testList "Data Handler" [
  let withSetup f ():unit =
    let ``end`` = Matcher<string option, unit>()

    let handler = dataHandler' ``end``.Mock

    f (``end``, handler)

  yield! testFixture withSetup [
    "Should call end with map for info event", fun (``end``, handler) ->
      handler (toJson """{ "ACTION": "info" }""")
      ``end`` <?> Some("{}")

    "Should call end for add event", fun (``end``, handler) ->
       handler addObj
       ``end`` <?> None;

    "Should call end for remove event", fun (``end``, handler) ->
      handler removeObj
      ``end`` <?> None;

    "Should end on a bad match", fun (``end``, handler) ->
      expect.assertions 2
      try
        handler (toJson """{}""")
      with
        | ex ->
          ``end`` <?> None
          expect.Invoke(ex.Message).toEqual("Handler got a bad match")

  ]
]
