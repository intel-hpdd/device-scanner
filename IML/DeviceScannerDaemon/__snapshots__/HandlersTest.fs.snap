// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Handler Should add dataset property then export then import 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"0\\"}}},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add dataset property then export then import 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"EXPORTED\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"0\\"}}},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add dataset property then export then import 3`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"0\\"}}},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add dataset property then export then import 4`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add multiple pool properties then add two datasets with multiple properties then export then import 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"0\\",\\"readonly\\":\\"1\\"}},\\"69\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/backup\\",\\"DATASET_UID\\":\\"69\\",\\"PROPERTIES\\":{\\"canmount\\":\\"1\\",\\"readonly\\":\\"0\\"}}},\\"PROPERTIES\\":{\\"failmode\\":\\"1\\",\\"multihost\\":\\"0\\"}}}}"`;

exports[`Data Handler Should add multiple pool properties then add two datasets with multiple properties then export then import 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"EXPORTED\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"0\\",\\"readonly\\":\\"1\\"}},\\"69\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/backup\\",\\"DATASET_UID\\":\\"69\\",\\"PROPERTIES\\":{\\"canmount\\":\\"1\\",\\"readonly\\":\\"0\\"}}},\\"PROPERTIES\\":{\\"failmode\\":\\"1\\",\\"multihost\\":\\"0\\"}}}}"`;

exports[`Data Handler Should add multiple pool properties then add two datasets with multiple properties then export then import 3`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"0\\",\\"readonly\\":\\"1\\"}},\\"69\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/backup\\",\\"DATASET_UID\\":\\"69\\",\\"PROPERTIES\\":{\\"canmount\\":\\"1\\",\\"readonly\\":\\"0\\"}}},\\"PROPERTIES\\":{\\"failmode\\":\\"1\\",\\"multihost\\":\\"0\\"}}}}"`;

exports[`Data Handler Should add multiple pool properties then add two datasets with multiple properties then export then import 4`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{\\"canmount\\":\\"1\\",\\"readonly\\":\\"1\\"}},\\"69\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/backup\\",\\"DATASET_UID\\":\\"69\\",\\"PROPERTIES\\":{\\"canmount\\":\\"1\\",\\"readonly\\":\\"0\\"}}},\\"PROPERTIES\\":{\\"failmode\\":\\"1\\",\\"multihost\\":\\"1\\"}}}}"`;

exports[`Data Handler Should add multiple pool properties then add two datasets with multiple properties then export then import 5`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{}}"`;

exports[`Data Handler Should add pool property then export then import 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{\\"multihost\\":\\"0\\"}}}}"`;

exports[`Data Handler Should add pool property then export then import 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"EXPORTED\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{\\"multihost\\":\\"0\\"}}}}"`;

exports[`Data Handler Should add pool property then export then import 3`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{\\"multihost\\":\\"0\\"}}}}"`;

exports[`Data Handler Should add pool property then export then import 4`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{}}"`;

exports[`Data Handler Should add then remove a device path 1`] = `"{\\"BLOCK_DEVICES\\":{\\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\":{\\"MAJOR\\":\\"8\\",\\"MINOR\\":\\"17\\",\\"PATHS\\":[\\"/dev/sdb1\\",\\"/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1\\",\\"/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1\\"],\\"DEVNAME\\":\\"/dev/sdb1\\",\\"DEVPATH\\":\\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\",\\"DEVTYPE\\":\\"partition\\",\\"ID_VENDOR\\":null,\\"ID_MODEL\\":\\"VBOX_HARDDISK\\",\\"ID_SERIAL\\":\\"VBOX_HARDDISK_VB304a0a0f-15e93f07\\",\\"ID_FS_TYPE\\":\\"LVM2_member\\",\\"ID_PART_ENTRY_NUMBER\\":1,\\"IML_SIZE\\":\\"81784832\\",\\"IML_SCSI_80\\":\\"80\\",\\"IML_SCSI_83\\":\\"83\\",\\"IML_IS_RO\\":true,\\"IML_DM_SLAVE_MMS\\":null,\\"IML_DM_VG_SIZE\\":null,\\"DM_MULTIPATH_DEVICE_PATH\\":true,\\"DM_LV_NAME\\":\\"swap\\",\\"DM_VG_NAME\\":\\"centos\\",\\"DM_UUID\\":\\"LVM-pV8TgNKMJVNrolJgMhVwg4CAeFFAIMC83IU1hvimWWlvmd5xQddtMIqRtjwOuKTz\\"}},\\"ZFSPOOLS\\":{}}"`;

exports[`Data Handler Should add then remove a device path 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{}}"`;

exports[`Data Handler Should add then remove a zdataset 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{}}},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add then remove a zdataset 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add then remove a zpool 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should add then remove a zpool 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{}}"`;

exports[`Data Handler Should call end with map for info event 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{}}"`;

exports[`Data Handler Should end on a bad match 1`] = `"Handler got a bad match"`;

exports[`Data Handler Should export then import zpool with datasets 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"EXPORTED\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{}}},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should export then import zpool with datasets 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{\\"11\\":{\\"POOL_UID\\":\\"0x2D28F440E514007F\\",\\"DATASET_NAME\\":\\"testPool1/home\\",\\"DATASET_UID\\":\\"11\\",\\"PROPERTIES\\":{}}},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should export then import zpool with datasets 3`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"EXPORTED\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should export then import zpool with datasets 4`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should fail when adding a dataset to non-existent pool 1`] = `"Pool to update dataset on is missing! \\"0x2D28F440E514007F\\""`;

exports[`Data Handler Should fail when adding a property to non-existent dataset 1`] = `"Dataset to update property on is missing! \\"11\\" (pool \\"0x2D28F440E514007F\\")"`;

exports[`Data Handler Should fail when adding a property to non-existent dataset on a non-existent pool 1`] = `"Pool to update dataset property on is missing! \\"0x2D28F440E514007F\\""`;

exports[`Data Handler Should fail when adding a property to non-existent pool 1`] = `"Pool to update property on is missing! \\"0x2D28F440E514007F\\""`;

exports[`Data Handler Should import then export then import a zpool 1`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should import then export then import a zpool 2`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"EXPORTED\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;

exports[`Data Handler Should import then export then import a zpool 3`] = `"{\\"BLOCK_DEVICES\\":{},\\"ZFSPOOLS\\":{\\"0x2D28F440E514007F\\":{\\"NAME\\":\\"testPool1\\",\\"UID\\":\\"0x2D28F440E514007F\\",\\"STATE_STR\\":\\"ACTIVE\\",\\"PATH\\":\\"testPool1\\",\\"DATASETS\\":{},\\"PROPERTIES\\":{}}}}"`;
